### PR TITLE
Open by serial

### PR DESCRIPTION
**Pull request by Novakov**

### CODE REVIEW

File: src/uffs/uffs_fs.c [hunk 1]
```diff
@@ -557,6 +557,71 @@ URET uffs_OpenObjectEx(uffs_Object *obj, uffs_Device *dev,
...
+		obj->err = UEINVAL;
+		goto ext;
+	}
+
+	obj->dev = dev;
+	obj->type = (oflag & UO_DIR ? UFFS_TYPE_DIR : UFFS_TYPE_FILE);
+
+	uffs_ObjectDevLock(obj);
+
+	if (obj->type == UFFS_TYPE_DIR) {
```
Comment (mandrek44) [inline src/uffs/uffs_fs.c:604]: In `uffs_OpenObejctEx`, when you try to open root directory (with serial = `ROOT_DIR_SERIAL`) then `obj->node` is not propagated. See line 499 in this file

File: src/uffs/uffs_fs.c [hunk 2]
```diff
@@ -557,6 +557,71 @@ URET uffs_OpenObjectEx(uffs_Object *obj, uffs_Device *dev,
...
+	obj->oflag = oflag;
+	obj->parent = GET_OBJ_NODE_FATHER(obj);
+	obj->pos = 0;
+	obj->name = 0;
+	obj->name_len = 0;
+
+	obj->sum = 0;
+	obj->head_pages = obj->dev->attr->pages_per_block - 1;
+
+	if ((obj->oflag & (UO_CREATE | UO_EXCL)) == (UO_CREATE | UO_EXCL)){
```
Comment (mandrek44) [inline src/uffs/uffs_fs.c]: We check only UO_CREATE flag when obj exists, but we return `UENOENT` when object doesn't exist even if UO_CREATE is set - this is inconsistency. 

For me using this method on non-existing object and with UO_CREATE flag is wrong. I suggest we return UEINVAL error if UO_CREATE flag is set (with some log entry)

File: src/uffs/uffs_fs.c [hunk 3]
```diff
@@ -557,6 +557,71 @@ URET uffs_OpenObjectEx(uffs_Object *obj, uffs_Device *dev,
...
+		obj->err = UEEXIST;
+		goto ext_1;
+	}
+
+	obj->serial = GET_OBJ_NODE_SERIAL(obj);
+	obj->open_succ = U_TRUE;
+
+	if (obj->oflag & UO_TRUNC)
```
Comment (mandrek44) [inline src/uffs/uffs_fs.c]: I know that original code didn't have brackets for this `if` but maybe lets fix that in here.

File: src/uffs/uffs_fs.c (hunk 1 above)
Comment (Novakov) [reply src/uffs/uffs_fs.c:604]: done

File: src/uffs/uffs_fs.c (hunk 2 above)
Comment (Novakov) [reply src/uffs/uffs_fs.c]: done

**Review by mandrek44 — APPROVED** [review]

**Pull request merged**